src/x86_64/page.c

```diff
@@ -423,6 +423,32 @@ void zero_mapped_pages(u64 vaddr, u64 length)
     traverse_entries(vaddr, length, stack_closure(zero_page));
 }
 
+static CLOSURE_2_3(is_page_mapped, boolean, u64, u8 *, int, u64, u64 *);
+boolean is_page_mapped(u64 baseaddr, u8 * vec, int level, u64 addr, u64 * entry)
+{
+    u64 e = *entry;
+    int pgoff, i;
+
+    if (entry_is_present(e)) {
+        pgoff = (addr - baseaddr) >> PAGELOG;
+
+        if (entry_is_fat(level, e)) {
+            /* whole level is mapped */
+            for (i = 0; i < 512; i++)
+                vec[pgoff + i] = 1;
+        } else if (entry_is_pte(level, e)) {
+            vec[pgoff] = 1;
+        }
+    }
+
+    return true;
+}
+
+void mincore_pages(u64 vaddr, u64 length, u8 *vec)
```

Comment:
In general, we've tried to keep the unix-specific stuff confined to the unix world (i.e. src/unix) and the low-level kernel stuff as generic as possible. I think in this case it would make sense to make traverse_entries() global (and make the name more descriptive, e.g. traverse_ptes or traverse_page_tables) and then move is_page_mapped back to mmap.c and invoke the traverse directly from within mincore().

Reply:
OK, makes sense